core/unix/module_elf.c

```diff
@@ -298,13 +360,23 @@ module_fill_os_data(ELF_PROGRAM_HEADER_TYPE *prog_hdr, /* PT_DYNAMIC entry */
             if (soname_index != -1 && dynstr != NULL) {
                 *soname = dynstr + soname_index;
 
-                /* sanity check soname location */
-                if ((app_pc)*soname < base || (app_pc)*soname > base + sz) {
+                /* Sanity check soname location using base address and size
+                 * depending on if virtual or file offset has been used. See
+                 * i#3385 comment above.
+                 */
+                app_pc base_variant;
+                size_t size_variant;
+                if (at_map == 1) {
+                    base_variant = out_data->source_file_map;
```

Comment:
It looks like this mmap base is only directly used within this single function.  The `*soname` is passed through some callers and eventually `dr_strup`ed.  Maybe we could change the interface to call `dr_strdup` here when `soname` != NULL?  If a caller doesn't want it, they don't pass it in; if they want it, they now have to own it.  Then we can change the map base to just be a local variable and we can unmap the file in this function.

Reply:
I'll implement this shortly. Just to be clear, we want:
- `module_fill_os_data()` to be responsible for allocating memory for `soname` using `dr_strdup` and the caller to be responsible for freeing using `dr_strfree()`?
- If the caller is not interested in `soname` it passes NULL so `module_fill_os_data()` doesn't do anything with `soname`?

Which one of the heap accounting types is best for `dr_strdup()`? `ACCT_OTHER`? It's not clear from the code what each one means. It looks like it depends on which part of the codebase is doing the calling.

Reply:
Yes, that is my proposal.  ACCT_VMAREAS.  I would say either document that the caller needs to free with ACCT_VMAREAS too, or have the caller pass it in like is done with os_module_area_init() to ensure the free has the same value.

Reply:
I implemented what I thought was going to be a simple change.
drrun with no clients only the subject binary works fine.
But running with a client resulted in strange segfaults happening quite early in the run.

In order to simplify debugging, I reduced the change to a minimum.
It appears that the mere presence of `dr_strdup()` in `module_fill_os_data()` is enough to cause the failure!

```
diff --git a/core/unix/module_elf.c b/core/unix/module_elf.c
index db9e6e25..1f12bdee 100644
--- a/core/unix/module_elf.c
+++ b/core/unix/module_elf.c
@@ -258,6 +258,9 @@ module_fill_os_data(ELF_PROGRAM_HEADER_TYPE *prog_hdr, /* PT_DYNAMIC entry */
     dcontext_t *dcontext = get_thread_private_dcontext();
     /* i#489, DT_SONAME is optional, init soname to NULL first */
     *soname = NULL;
+    const char *s = dr_strdup("what?!" HEAPACCT(ACCT_VMAREAS));
+    SYSLOG_INTERNAL_INFO("DEBUG: s=[%s]", s);
+    dr_strfree(s HEAPACCT(ACCT_VMAREAS));
 #ifdef ANDROID
     /* On Android only the first segment is mapped in and .dynamic is not
      * accessible.  We try to avoid the cost of the fault.
@@ -366,6 +369,7 @@ module_fill_os_data(ELF_PROGRAM_HEADER_TYPE *prog_hdr, /* PT_DYNAMIC entry */
                 if (*soname != NULL && strlen(*soname) == -1) {
                     ASSERT_NOT_REACHED();
                 }
+            SYSLOG_INTERNAL_INFO("DEBUG: *soname=[%s]", *soname);
             }
             if (out_data != NULL) {
                 /* we put module_hashtab_init here since it should always be called
lines 1-22/22 (END)

```
This results in the assert firing in `TRY_EXCEPT_ALLOW_NO_DCONTEXT`:

```
$ $DRBUILD/bin64/drrun -c $DRBUILD/api/bin/libinscount.so -- ./hw                                                                      
<DEBUG: s=[what?!]>
<Application /path/to/hw (6268).  Internal Error: DynamoRIO debug check failure: /path/to/i3385-elf-dynstr-fix/dynamorio/core/unix/module_elf.c:385 try__except != NULL
(Error occurred @0 frags)
version 7.1.17970, custom build
-no_dynamic_options -no_deadlock_avoidance -client_lib '/path/to/i3385-elf-dynstr-fix/dynamorio/build/api/bin/libinscount.so;0;;/path/to/i3385-elf-dynstr-fix/dynamorio/build/api/bin/libinscount.so;0;' -code_api -stack_size 56K -signal_stack_size 32K -max_elide_jmp 0 -max_elide_call 0                                 
0x0000ffffeb5ff650 0x00000000710bd120
0x0000ffffeb5ff7f0 0x0000000071325fbc
. . .
0x0000ffffeb601cf0 0x00000000712e99a4>
$

```

```
#define TRY_EXCEPT_ALLOW_NO_DCONTEXT(dcontext, try_statement, except_statement) \
    do {                                                                        \
        . . .                                                                   \
        ASSERT(try__except != NULL);     <---                                   \
        TRY(try__except, try_statement, EXCEPT(try__except, except_statement)); \
    } while (0)

```
To simplify the change even further and isolate the cause, I removed the try/except in `module_fill_os_data()`. This results in a segfault:

```
$ $DRBUILD/bin64/drrun -c $DRBUILD/api/bin/libinscount.so -- ./hw 
<DEBUG: s=[what?!]>
<DEBUG: *soname=[libinscount.so]>
<DEBUG: s=[what?!]>
<DEBUG: *soname=[libinscount.so]>
Segmentation fault (core dumped)
$

```
GDB shows it crashes in `LOCK_VECTOR` because the `vm_area_vector_t*` is 0:

```
(gdb) r -c $DRBUILD/api/bin/libinscount.so -- ./hw
Starting program: /path/to/i3385-elf-dynstr-fix/dynamorio/build/bin64/drrun -c $DRBUILD/api/bin/libinscount.so -- ./hw
Missing separate debuginfo for /lib/ld-linux-aarch64.so.1
Try: zypper install -C "debuginfo(build-id)=e8104675ba94d7c698d02558d81423b4fe5bff11"
Missing separate debuginfo for /lib64/libc.so.6
Try: zypper install -C "debuginfo(build-id)=9a94d7a23a8802dd0f216f3a3cea0eb29d703de0"
process 6944 is executing new program: /path/to/i3385-elf-dynstr-fix/dynamorio/build/lib64/debug/libdynamorio.so
<DEBUG: s=[what?!]>
<DEBUG: *soname=[libinscount.so]>
<DEBUG: s=[what?!]>
<DEBUG: *soname=[libinscount.so]>

Program received signal SIGSEGV, Segmentation fault.
0x000000007117b4b0 in vmvector_add (v=0x0, start=0x72000000 "\177ELF\002\001\001\003", end=0x72005000 "", data=0x38181880)
    at /path/to/i3385-elf-dynstr-fix/dynamorio/core/vmareas.c:1891
1891        LOCK_VECTOR(v, release_lock, write);
(gdb) bt
#0  0x000000007117b4b0 in vmvector_add (v=0x0, start=0x72000000 "\177ELF\002\001\001\003", end=0x72005000 "", data=0x38181880)
    at /path/to/i3385-elf-dynstr-fix/dynamorio/core/vmareas.c:1891
#1  0x000000007131d56c in privload_add_areas (privmod=0x38181880) at /path/to/i3385-elf-dynstr-fix/dynamorio/core/unix/loader.c:374
#2  0x00000000711c99d0 in privload_load_finalize (privmod=0x38181880) at /path/to/i3385-elf-dynstr-fix/dynamorio/core/loader_shared.c:689
#3  0x00000000711c9224 in privload_load (filename=0xffffffffbdd0 "/path/to/i3385-elf-dynstr-fix/dynamorio/build/api/bin/libinscount.so", dependent=0x0, 
    client=true) at /path/to/i3385-elf-dynstr-fix/dynamorio/core/loader_shared.c:571
#4  0x000000007131e274 in privload_locate_and_load (impname=0xffffffffccd0 "/path/to/i3385-elf-dynstr-fix/dynamorio/build/api/bin/libinscount.so", 
    dependent=0x0, reachable=true) at /path/to/i3385-elf-dynstr-fix/dynamorio/core/unix/loader.c:651
#5  0x000000007131e2d0 in privload_load_private_library (name=0xffffffffccd0 "/path/to/i3385-elf-dynstr-fix/dynamorio/build/api/bin/libinscount.so", 
    reachable=true) at /path/to/i3385-elf-dynstr-fix/dynamorio/core/unix/loader.c:663
#6  0x00000000711c7e00 in locate_and_load_private_library (name=0xffffffffccd0 "/path/to/i3385-elf-dynstr-fix/dynamorio/build/api/bin/libinscount.so", 
    reachable=true) at /path/to/i3385-elf-dynstr-fix/dynamorio/core/loader_shared.c:232
#7  0x00000000712f08fc in load_shared_library (name=0xffffffffccd0 "/path/to/i3385-elf-dynstr-fix/dynamorio/build/api/bin/libinscount.so", reachable=true)
    at /path/to/i3385-elf-dynstr-fix/dynamorio/core/unix/os.c:3784
#8  0x00000000711d2f20 in add_client_lib (path=0xffffffffccd0 "/path/to/i3385-elf-dynstr-fix/dynamorio/build/api/bin/libinscount.so", 
    id_str=0xffffffffcd28 "0", options=0xffffffffcd2a "") at /path/to/i3385-elf-dynstr-fix/dynamorio/core/lib/instrument.c:543
#9  0x00000000711d35d8 in instrument_load_client_libs () at /path/to/i3385-elf-dynstr-fix/dynamorio/core/lib/instrument.c:662
#10 0x000000007102a33c in dynamorio_app_init () at /path/to/i3385-elf-dynstr-fix/dynamorio/core/dynamo.c:493
#11 0x0000000071321208 in privload_early_inject (sp=0xffffffffe7d0, old_libdr_base=0xffffbf2c6000 <error: Cannot access memory at address 0xffffbf2c6000>, 
    old_libdr_size=4435968) at /path/to/i3385-elf-dynstr-fix/dynamorio/core/unix/loader.c:1947
#12 0x00000000712e99a4 in _start () at /path/to/i3385-elf-dynstr-fix/dynamorio/core/arch/aarchxx/aarchxx.asm:66
Backtrace stopped: previous frame identical to this frame (corrupt stack?)
(gdb)

```

- Given the memory management work dr_strdup() is doing under the covers, are there limits to how early it can be called?
- Are there some memory locking/initialisation issues which we need to handle?

I notice there is one other `dr_strdup()` in `os_module_update_dynamic_info()` but I've never seen that being called.
I tried some of the other `ACCT_` heap accounting types but that makes no difference.

Reply:
Note that client libs are setup in two parts, and the early part, instrument_load_client_libs(), is called right before d_r_heap_init().  However, with #2558 breaking the vmm setup dependence, I believe we can move instrument_load_client_libs() down below heap init, which should solve this problem.

Reply:
> I believe we can move instrument_load_client_libs() down below heap init, which should solve this problem.                                                       

Yes, the problem was solved by moving `instrument_load_client_libs()` after `d_r_heap_init()`.                                                                                       
I implemented the change to `soname`'s memory allocation and freeing as discussed so that we could free the mapping in `module_fill_os_data()` and still have a copy of `soname`.      
However, this caused test regressions:

```
The following tests FAILED:
         87 - code_api|client.modules (Failed)
        118 - code_api|client.drwrap-test (Failed)
        120 - code_api|client.drsyms-test (Failed)
        132 - code_api|sample.wrap (Failed)
```

All except `client.modules` were caused by unmapping in `module_fill_os_data()`.
For example, `sample.wrap` crashes when it tries to access `strtab` using the mappings `dynstr` in `elf_sym_matches()`.                                                                  
This can be fixed by unmapping in `os_loader_exit()` rather than `module_fill_os_data()`.

I've committed this fix leaving the `soname` memory allocation change in for now, at least until `client.modules` is fixed.                                                          

This does not fix client.modules, still debugging that.
It segfaults in` privload_lookup()` because `char *name` is bad.

Reply:
> All except client.modules were caused by unmapping in module_fill_os_data().
> For example, sample.wrap crashes when it tries to access strtab using the mappings dynstr in elf_sym_matches().
> This can be fixed by unmapping in os_loader_exit() rather than module_fill_os_data().

Are the failures because we're storing the `strtab` file offset instead of vritual address?  And are thus reliant on the flat mapping?  It doesn't seem ideal for DR to keep a separate mapping of every single app library (we have some some cases of thousands of libraries, yes thousands) for their entire lifetimes.  I'm not saying I have a simple solution though.  Are these bad accesses much later during execution, after the library's segments are mapped?  I.e., could this be solved by switching to the virtual address for steady state?  If these are in clients maybe they're only after the 1st execution.  Maybe that could be a trigger point to unload the flat mapping and switch to virtual addresses?  Getting more complex though.  I'm just afraid of filling up a 32-bit address space of a large app that on its own nearly uses the whole space, by doubling all the libraries.

Reply:
I'm also confused about why changing just soname ends up causing regressions in accessing other data?

Reply:
> I'm also confused about why changing just soname ends up causing regressions in accessing other data?

Extracting `soname` is based on establishing `dynstr`, which is either file or virtual offset, depending on `at_map`. It's the value of `dynstr` (and hence `DT_STRTAB`) which is the cause of the regressions.

Reply:
> Are these bad accesses much later during execution, after the library's segments are mapped?

I'm guessing it is because the test clients are trying to access unmapped memory and I assume that by the time client code is executed, all segments are mapped.

> i.e., could this be solved by switching to the virtual address for steady state?

Which event(s) signal all segments are loaded and we are at steady state?

Reply:
The only thing we have today is the first execution of a module.  See #884 where we moved the client's module load event to the first execution.

Could we do this:
1) Unmap the just-for-DR flat mmap right away (`module_fill_os_data`) and strdup the soname.
At that time, clear dynstr or mark as invalid in some way to avoid DR's internal code from trying to access the unmapped file.  (Also check for any other fields like this: is it just dynstr?)
2) On first execution of a module, look up the vaddr for dynstr and set it.

Unfortunately it's a little complex -- but I think all the prior approaches we talked about were even more so.